### ISSUE TITLE
Client generated IDs/extensibility?

### ISSUE THREAD

**Issue by jackrusher** [Unrepl/unrepl#8]
If I'm reading the current protocol documentation properly, the client is meant to send only a bare form for evaluation. This strikes me as worrisome for future development, as there's no way to add request metadata to client communication.

A short term motivating example for this would be to add a client-generated ID to each evaluation request that would then be included in any results returned by the server for that request. This method is more robust than using asynchronously-arriving server generated IDs for this purpose.

(The current mechanism relies on request order to assign each result to a specific request, which means that if there's a problem with any one request on the client side -- an emacs user `C-g`s before the local request queue state machine is updated, for example -- there's no principled way to recover that total ordering. On the other hand, If the client were to generate and send a UUID for each request it could simply ignore UUIDs that aren't in the local UUID->request lookup table, and thus recover from transient client failures of this type.)

In short, please consider using a map expressed as EDN for client-side communication, something like this:
```
{:form (+ 1 1)
 :uuid "xxx"}
```
... noting that this also opens the possibility of further extension points in the future, as we learn more about what is desirable in this context.

**Comment by cgrand** [comment]
I understand the concern but after some earlier iterations I came to the conclusion that the input stream must be left untouched. 

The `:echo` messages are meant to help the client to map outputs (of any kind) to input. They give position of the input both in the input stream (as offset from the start) and in source files (as line/col).
